documentation/design_docs/design/user-resource.adoc

```diff
@@ -0,0 +1,103 @@
+== why
+
+Tenants currently need to create users and assign permissions using the Keycloak console or API when
+using the standard address space.
+
+However, usage of Keycloak is an implementation detail, and also a very crude interface.
+
+This proposal is for creating users and managing user permissions using a custom API part of the api
+server. 
+
+== what/how
+
+A new api group 'auth.enmasse.io' will be created. It will support the following operations:
+
+To create a user (oc create -f user1.yaml):
+```
+POST /apis/auth.enmasse.io/v1alpha1/namespaces/myproject/users
+```
+
+To read a user (oc get myspace.user1):
+```
+GET /apis/auth.enmasse.io/v1alpha1/namespaces/myproject/users/myspace.user1
+```
+
+To update a user (oc replace -f user1.yaml):
+
+```
+PUT /apis/auth.enmasse.io/v1alpha1/namespaces/myproject/users/myspace.user1
+```
+
+To remove a user (oc delete enmasse-user myspace.user1)
+
+```
+DELETE /apis/auth.enmasse.io/v1alpha1/namespaces/myproject/users/myspace.user1
+```
+
+In the same way as addresses, users are scoped by namespace _and_ address space and resource names
+must include the address space name and username separated by a dot ('.').
+
+A user may authentication using passwords or a federated identity. In the case of authentication
+using a password, the password is provided upon creation on the user.  Subsequent reads of this
+user will not be able to see the password, and so applications need to 'remember' the password
+using secrets or other mechanisms.
+
+```
+apiVersion: auth.enmasse.io/v1alpha1
+kind: EnMasseUser
+metadata:
+    name: myspace.user1
+spec:
+    authentication:
+      type: password
+      value: mypassword
+    authorization:
+    - operations: [ "create, "delete", "send", "receive", "view" ]
+      addresses: [ "queue1", "queue2", "*", "queue*" ]
+```
+
+Users can also be authenticated using a federated identity where providers such as OpenShift can be
+supported.
+
+```
+apiVersion: auth.enmasse.io/v1alpha1
+kind: EnMasseUser
+metadata:
+    name: myspace.user2
+spec:
+    authentication:
+      type: federated
+      provider: openshift
+      openShiftUserId: openshift user id
+      openShiftUserName: openshift user name
+    authorization:
+    - operations: [ "create, "delete", "send", "receive", "view", "*" ]
+      addresses: [ "queue1", "queue2", "*", "queue*" ]
+```
+
```

Comment:
I think it would be better to have the authorization keyed by the address/address pattern and then have the operations allowed on that address defined per address

```
authorization:
  - addresses: ["queue1", "queue2"]
    operations: ["send", "receive"]
  - addresses: ["*"]
    operations: ["view"]
  ```

Reply:
Not sure i understand completely. There is no key here, just arrays. There are not connection between the order of items in the operations or the addresses array.